src/mbgl/programs/programs.hpp

```diff
@@ -1,82 +1,54 @@
-          line(context, programParameters),
-          lineGradient(context, programParameters),
-          lineSDF(context, programParameters),
-          linePattern(context, programParameters),
-          raster(context, programParameters),
-          symbolIcon(context, programParameters),
-          symbolIconSDF(context, programParameters),
-          symbolGlyph(context, programParameters),
-          debug(context, programParameters),
-          collisionBox(context, programParameters),
-          collisionCircle(context, programParameters),
-          clippingMask(context, programParameters) {
-    }
+    Programs(gl::Context&, const ProgramParameters&);
+    ~Programs();
 
-    BackgroundProgram background;
-    BackgroundPatternProgram backgroundPattern;
-    ProgramMap<CircleProgram> circle;
-    ExtrusionTextureProgram extrusionTexture;
-    ProgramMap<FillProgram> fill;
-    ProgramMap<FillExtrusionProgram> fillExtrusion;
-    ProgramMap<FillExtrusionPatternProgram> fillExtrusionPattern;
-    ProgramMap<FillPatternProgram> fillPattern;
-    ProgramMap<FillOutlineProgram> fillOutline;
-    ProgramMap<FillOutlinePatternProgram> fillOutlinePattern;
-    ProgramMap<HeatmapProgram> heatmap;
-    HeatmapTextureProgram heatmapTexture;
-    HillshadeProgram hillshade;
-    HillshadePrepareProgram hillshadePrepare;
-    ProgramMap<LineProgram> line;
-    ProgramMap<LineGradientProgram> lineGradient;
-    ProgramMap<LineSDFProgram> lineSDF;
-    ProgramMap<LinePatternProgram> linePattern;
-    RasterProgram raster;
-    ProgramMap<SymbolIconProgram> symbolIcon;
-    ProgramMap<SymbolSDFIconProgram> symbolIconSDF;
-    ProgramMap<SymbolSDFTextProgram> symbolGlyph;
+    BackgroundLayerPrograms& getBackgroundLayerPrograms() noexcept;
+    RasterLayerPrograms& getRasterLayerPrograms() noexcept;
+    HeatmapLayerPrograms& getHeatmapLayerPrograms() noexcept;
+    CircleLayerPrograms& getCircleLayerPrograms() noexcept;
+    HillshadeLayerPrograms& getHillshadeLayerPrograms() noexcept;
+    FillLayerPrograms& getFillLayerPrograms() noexcept;
+    FillExtrusionLayerPrograms& getFillExtrusionLayerPrograms() noexcept;
+    LineLayerPrograms& getLineLayerPrograms() noexcept;
+    SymbolLayerPrograms& getSymbolLayerPrograms() noexcept;
 
     DebugProgram debug;
-    CollisionBoxProgram collisionBox;
-    CollisionCircleProgram collisionCircle;
     ClippingMaskProgram clippingMask;
+
+private:
+    std::unique_ptr<LayerTypePrograms> backgroundPrograms;
+    std::unique_ptr<LayerTypePrograms> circlePrograms;
+    std::unique_ptr<LayerTypePrograms> rasterPrograms;
+    std::unique_ptr<LayerTypePrograms> heatmapPrograms;
+    std::unique_ptr<LayerTypePrograms> hillshadePrograms;
+    std::unique_ptr<LayerTypePrograms> fillPrograms;
+    std::unique_ptr<LayerTypePrograms> fillExtrusionPrograms;
+    std::unique_ptr<LayerTypePrograms> linePrograms;
+    std::unique_ptr<LayerTypePrograms> symbolPrograms;
```

Comment:
Why do we need an abstract base class here?

Reply:
`unique_ptr` requires a destructor defined, so pointers to the concrete classes would prevent from LTO removing them from the binaries.